src/tools/aclip/aclip.1
.\" groff -man -Tascii aclip.1
.TH arcan 1 "February 2017" aclip "User manual"
.SH NAME
aclip \- command line interface to Arcan clipboards
.SH SYNOPSIS
.B aclip [OPTIONS]

.SH DESCRIPTION
This tool is used to send input to- and/or monitor- an arcan connection point
that accepts clipboards as a primary connection, which is an indicator of
a clipboard manager. It is up to the running set of scripts that controls
the connection point to determine if this is permitted or not, but
the 'safe default' is rejecting such connections. See 'Limitations' below.

.SH OPTIONS
.IP "\fB\-i, \-\-in\fR"
Read valid UTF-8 input from standard input and forward to the clipboard.

.IP "\fB\-I, \-\-in-data \fIdatastr\fR"
Forward the contents of the datastr argument to the clipboard.

.IP "\fB\-o, \-\-out\fR"
Enable output mode, sending incoming clipboard entries (by default) to stdout.

.IP "\fB\-e \-\-exec \fIexecstr\fR"
For each new clipboard entry, run (popen) execstr in a subshell, where the
entry contents will be available on stdin.

.IP "\fB\-p, \-\-separator \fIsepstr\fR"
(in -o mode) Append sepstr argument after each incoming clipboard entry.

.IP "\fB\-l, \-\-loop\fR \fIcount\fR"
(in -o mode) Set a limit of new entries that should be received from the
clipboard. If count is set to '0', the program will loop indefinitely.

.IP "\fQ\-s, \-\-silent\fR"
Close stdout and fork into background (so best use with -e or -i/-I, not just -o)

.IP "\fB\-d, \-\-display\fR \fIconnpath\fR"
By default, the ARCAN_CONNPATH environment is used in order to find an
arcan instance to connect to. This argument overrides the environment
with the connpath argument.

.IP "\fB\-h, \-\-help\fR"
Show a brief summary of commands and usage.

.SH EXAMPLES

.PP
This examples places 'hi there' on the clipboard.

.B echo -n 'hi there' | aclip -p durden -i

.PP
Keep the X selection buffer synched to the latest received clipboard entry.

.B aclip -o -l 0 -s -e xclip

.PP
Terminate after three completed clipboard entries. each new entry, spawn the
"tee" command that appends to a log with the custom '||' string to separate
between elements.

.B aclip -l 0 -o -p "||" -e "tee -a >> log"

.SH LIMITATIONS
The appl that arcan is running must expose a connection point, and explicitly
allow clipboard and/or clipboard-paste primary segments to connect. This is
usually not added by default due to the security implications it might have.

Additional constraints may also have been placed on this connection point.
One common constraint is to only allow one active monitor or input connection.
Other constraints may be read- or write- only access or ignoring write-
operations that is longer than a certain cutoff value.

Consult the documentation for the appl you are using to get more information
on what applies to your particular scenario.

All inputs and outputs need to be valid UTF-8 or they will likely be silently
filtered. The tool is still young and few of the clipboard features offered by
Arcan are exposed yet, this means that right now, only adding text to a
clipboard, and monitoring the clipboard for entries actually work. Managing
binary blobs, audio and video streams will be added in the near future.

.SH ENVIRONMENT
.TP
.SM
\fBARCAN_CONNPATH\fR
Name of the connection point to use when connecting to arcan.
.TP
.SH COPYRIGHT
Copyright  ©  2017 Bjorn Stahl. 3-clause BSD licensed. This is free software:
you are free  to  change and redistribute it. There is NO WARRANTY,
to the extent permitted by law.

.SH AUTHOR
Bjorn Stahl <contact at arcan-fe dot com>


CHANGELOG.md
## 0.7.1
## Lua
 * Changed builtin 'print' function to emit a parser friendly output when not traced
 * Added monitor commands for backtrace, eval, locals, stepping and specialised dumps
 * Added breakpoint hooks for all transitions into VM execution
 * Dynamic / on-demand attaching (local,remote) debugger (see 7c7ba007c3)

## Terminal
 * Expose args values as 'tui.arguments' table

## Net
 * Several fixes to binary in/out transfers

## Shmif
 * Migrating more OS specific work to platform
 * Add timeout+liveness check for recvmsg for edge condition shutdown
 * Allow enqueueing TARGET\_COMMAND\_EXIT to terminate one segment from another thread
 * Factored sem handle use out of cont

## 0.7.0
## Core
 * Wired in rendertarget vobj export for hwenc, opt-in via target\_flags on rectgt
 * Add basic positional audio support

## Platform
 * posix/glob : add asynch form
 * egl-dri: add nvidia\_gbmbo_fix option to fix scanout allocation for (some) nvidia GPUs
 * egl-dri: fixes to CRTC picking logic

## Lua
 * add overloaded glob\_resource that can return an open\_nonblock table
 * add alttext slot for image\_tracetag
 * add audio\_position / audio\_listener for setting a video object as positional audio reference
 * add basic text\_surface for simplified text with a rendering path similar to tui windows
 * image\_access\_storage
 * add audio\_reconfigure for toggling hrtfs and switching between outputs
 * prohibit bytecode in system\_load

## Distribution
 * builtin/mouse opt-in rmb drag, unified theme handling, drop native_ path

## Shmif
 * add interop helper for arcan\_shmif\_bchunk\_resolve to help translate fd-local path
 * add a \_primary slot for ACCESSIBILITY
 * add an internal-catch segment for ACCESSIBILITY
 * default-route ACESSIBILITY internal-catch to map a TUI window feed through encode in OCR mode
 * add RHINT\_EMPTY to use SHMIF\_SIGVID for clocking without GPU transfers
 * fixes several C++ interop problems with header definition
 * drop VOBJ substructure, passing vector objects as BCHUNK is much less complex

## Net
 * IPv6 discovery controls added
 * add -c file for assigning lua scriptable command-line overrides
 * headless runner for arcan-net host appl can be access via ANET\_RUNNER env.
 * spawning server-side Lua runner if matching appl found, controls message routing
 * introduce rekeying command for forward secrecy, placeholder PQ step-up and resumption
 * add support for directory file store/retrieval

## Decode
 * tts now exposes more input labels (INC/DEC/SETRATE)
 * tts flush speech buffers on reset\_target
 * tts exposes endpause control
 * add 'spell' protocol for generating suggestions, dependency: hunspell

## Tui
 * Copy window widget re-added as a forced input-label toggle to all tui windows
 * Readline: better position suggestion on overflow
 * Readline: add option to supply completion popup title
 * Readline: add option to provide per-item help-text
 * Readline: fix padding error when stepping history to shorter items
 * Expose accessibility window as mappable type
 * Expose cursor state constrols
 * Tui: prevent cursor from moving out of screen bounds
 * Lua Bindings: add helper :tempfile, :tempdir, :mkdir, :funlink, :fmkdir
 * Lua Bindings: nbio fixes

## VRbridge
 * Merge in pending- OHMD Xreal Air/2/2Pro support

## 0.6.3
## Lua
 * inbound events now have a 'frame' tag for pairing with verbose-frame notification
 * util:random\_interval(low,high) added for CSPRNG non-biased interval RNG
 * util:random\_bytes(len) added for CSPRNG byte string
 * target_input long messages are now marked as multipart
 * convey tui/tpack state in resize events
 * target\_anchorhint added for informing clients about positioning and hierarchy
 * video\_displaymode expose eotf / coordinates for primaries and contents light levels
 * open\_nonblock can now adopt an existing iostream into a target vid
 * input\_remap\_translation overloaded form for serializing backing keymap
 * clockreq no longer forwarded for frameserver event handler
 * image\_metadata added for annotating a vid used when streaming/sharing/scanout HDR contents
 * functions creating files now apply a separate WRITEMASK (split from USERMASK)

## Core
 * respect border attribute in text rasteriser
 * added frame\_id to external events that pairs with shmif-SIGVID signals
 * optional tracy build for profiling (-DENABLE\_TRACY)
 * frameserver clock(stepframe) event handling extended (see shmif)

## Tui
 * nbio asynch type confusion fix (function becomes pcall:userdata)
 * nbio close without explicit flush
 * dock subtype exposed for bar/tray like integration
 * add 'swallow' request for new windows
 * cursor(caret) style and color override controls
 * readline: add completion style control (border)
 * readline: allow attached popup to replace in-band completion
 * readline: history traversal early-stop fixed
 * fixes for screencopy on proxy window
 * refactored deprecated tsm-screen away out of tui (1/2)

## Net
 * allow h264 passthrough, sidestepping local encode
 * re-add afsrv\_net, have it support directory, source and sink access modes.
 * split a12:// into a12:// and a12s:// ARCAN_CONNPATH with the former permitting connecting to unknowns
 * add interactive accept/reject/trust for arcan-net use
 * reserve 'outbound' domain for interactively added outbound keys
 * enforce domain separation for allowed pubkeys
 * directory mode now splits out into sandboxed worker processes
 * directory mode can now specify permissions per tag-group
 * directory mode dynamic push appl from client with permission
 * directory mode notification of appl updates and sources coming / leaving
 * directory mode registration of dynamic sinks
 * directory mode sourcing dynamic sinks (sink-inbound only)
 * directory mode source-sink pairing 'reachable source' and 'tunnel' transfer modes
 * changed packaging format for appls
 * net\_open("@stdin") can now be used to access a per-directory-appl messaging group

src/tools/aclip/aclip.c
/*
 * arcan-clipboard (aclip)
 * Copyright 2017-2018, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Description: Clipboard Integration tool
 */
#include <arcan_shmif.h>
#include <getopt.h>
#include <signal.h>
#include "utf8.c"

static char* separator = "";

static void paste(struct arcan_shmif_cont* out, char* msg, bool cont)
{
	arcan_event msgev = {
		.ext.kind = ARCAN_EVENT(MESSAGE)
	};

	if (!out || !out->vidp || !msg)
		return;

/* split into message size blocks, align backwards so full UTF8
 * seqs are passed, else server may dismiss or kill us */
	size_t len = strlen(msg);

	uint32_t state = 0, codepoint = 0;
	char* outs = msg;
	size_t maxlen = sizeof(msgev.ext.message.data) - 1;

	while (len > maxlen){
		size_t i, lastok = 0;
		state = 0;
		for (i = 0; i <= maxlen - 1; i++){
			if (UTF8_ACCEPT == utf8_decode(&state, &codepoint, (uint8_t)(msg[i])))
				lastok = i;

			if (i != lastok){
				if (0 == i)
					return;
			}
		}

		memcpy(msgev.ext.message.data, outs, lastok);
		msgev.ext.message.data[lastok] = '\0';
		len -= lastok;
		outs += lastok;
		if (len)
			msgev.ext.message.multipart = 1;
		else
			msgev.ext.message.multipart = cont;

		arcan_shmif_enqueue(out, &msgev);
	}

/* flush remaining */
	if (len){
		snprintf((char*)msgev.ext.message.data, maxlen, "%s", outs);
		msgev.ext.message.multipart = cont;
		arcan_shmif_enqueue(out, &msgev);
	}
}

static bool write_ev(arcan_event* ev, FILE* outf)
{
/* if it's a continuation, we can't write the separator yet, just find the
 * end of the buffer or where the string was truncated and terminated */
	if (ev->tgt.ioevs[0].iv != 0){
		size_t i = 0;
		for (; i < sizeof(ev->tgt.message); i++)
			if (ev->tgt.message[i] == '\0')
				break;
		if (i == 0)
			return false;
		if (i == sizeof(ev->tgt.message))
			i++;
		fwrite(ev->tgt.message, i, 1, outf);
		return false;
	}
	else{
		fprintf(outf, "%s%s", ev->tgt.message, separator);
		fflush(outf);
		return true;
	}
}

static int dispatch_event_outf(arcan_event* ev, FILE* outf)
{
	switch (ev->tgt.kind){
	case TARGET_COMMAND_MESSAGE:
		if (write_ev(ev, outf))
			return 1;
	break;
	case TARGET_COMMAND_NEWSEGMENT:
/* if we explicitly receive a new paste board due to some incoming stream */
	break;
	case TARGET_COMMAND_BCHUNK_IN:
	break;
	case TARGET_COMMAND_BCHUNK_OUT:
/* incoming binary blob, fd in ioevs[0].iv */
	break;
	case TARGET_COMMAND_STEPFRAME:
/* need to distinguish if we are an output segment or not, in the latter
 * case, we just received a raw video buffer to work with */
	break;
	case TARGET_COMMAND_EXIT:
		return -1;
	break;
	default:
	break;
	}
	return 0;
}

static int dispatch_event_outcmd(struct arcan_event* ev, const char* cmd)
{
	static FILE* outf;

	switch(ev->tgt.kind){
	case TARGET_COMMAND_MESSAGE:{
		if (!outf && !(outf = popen(cmd, "w")))
			return -1;
		write_ev(ev, outf);
		if (ev->tgt.ioevs[0].iv == 0){
			fclose(outf);
			outf = NULL;
			return 1;
		}
	}
	break;
	case TARGET_COMMAND_EXIT:
		return -1;
	break;
	default:
	break;
	}
	return 0;
}

static const struct option longopts[] = {
	{ "help",         no_argument,       NULL, 'h'},
	{ "in",           no_argument,       NULL, 'i'},
	{ "in-data",      required_argument, NULL, 'I'},
	{ "out",          no_argument,       NULL, 'o'},
	{ "exec",         required_argument, NULL, 'e'},
	{ "separator",    required_argument, NULL, 'p'},
	{ "loop",         required_argument, NULL, 'l'},
	{ "display",      required_argument, NULL, 'd'},
	{ "silent",       no_argument,       NULL, 's'},
	{ NULL,           no_argument,       NULL,  0 }
};

static void usage()
{
printf("Usage: aclip [-hioe:s:l:d:]\n"
"-h    \t--help         \tthis text\n"
"-i    \t--in           \tread/validate UTF-8 from standard input and copy to clipboard\n"
"-I arg\t--in-data arg  \tread/validate UTF-8 and copy to clipboard\n"
"-o    \t--out          \tflush received pastes to stdout\n"
"-e arg\t--exec arg     \tpopen [arg] on received pastes and flush to its stdin\n"
"-p arg\t--separator arg\ttreat [arg] as paste- separator separator (-l >= 0)\n"
"-l arg\t--loop arg     \texit after [arg] discrete paste operations (-0, never exit)\n"
"-s    \t--silent       \tclose stdout and fork into background\n"
"-d arg\t--display arg  \tuse [arg] as connection path istead of ARCAN_CONNPATH env\n"
);
}

int main(int argc, char** argv)
{
	int loop_counter = -1;

	if (argc == 1){
		usage();
		return EXIT_FAILURE;
	}

	bool use_stdin = false, use_stdout = false, silence = false;
	char* exec_cmd = NULL, (* copy_arg) = NULL;

	int ch;
	while((ch = getopt_long(argc, argv, "hisI:oe:p:l:d:", longopts, NULL)) >= 0)
	switch(ch){
	case 'h' : usage(); return EXIT_SUCCESS;
	case 'i' : use_stdin = true; break;
	case 'o' : use_stdout = true; break;
	case 's' : silence = true; break;
	case 'I' : {
		if (copy_arg)
			free(copy_arg);
		copy_arg = strdup(optarg);
	}
	break;
	case 'e' : {
		if (exec_cmd)
			free(exec_cmd);
		exec_cmd = strdup(optarg);
	}
	break;
	case 'd' : setenv("ARCAN_CONNPATH", optarg, 1); break;
	case 'l' : loop_counter = strtoul(optarg, NULL, 10); break;
	case 'p' : separator = strdup(optarg ? optarg : ""); break;
	default:
		break;
	}

	if (!use_stdin && !use_stdout && !copy_arg){
		fprintf(stderr, "neither [-i], [-I arg] nor [-o] specified.\n");
		usage();
		return EXIT_FAILURE;
	}

/* Something of an inconvenience right now, in order to get access to a proper
 * pasteboard (full A/V support) we need a non-output primary segment through
 * which that request can be pushed. If we get one, we can just switch the
 * struct that processes paste- output though. */
	struct arcan_shmif_cont con = arcan_shmif_open_ext(
		SHMIF_ACQUIRE_FATALFAIL, NULL, (struct shmif_open_ext){
		.type = SEGID_CLIPBOARD,
		.title = "",
		.ident = ""
	}, sizeof(struct shmif_open_ext));
	arcan_shmif_signal(&con, SHMIF_SIGVID);

	if (signal(SIGCHLD, SIG_IGN) == SIG_ERR){
		fprintf(stderr, "ign on sigchld failed\n");
		return EXIT_FAILURE;
	}

	if (silence){
		fclose(stdout);
		if (fork() != 0)
			return EXIT_SUCCESS;
	}

/*
 * Two little "gotcha's" here, one is that a lot of the event handler / resource
 * allocation schemes are actually deferred until the first transfer signal.
 * The other is a possible race: _drop pulls the 'dms' and the client is
 * considered dead. Whatever is pending on the event queue at the moment will
 * never be processed and our paste- message gets lost, so we want synch-
 * delivery here. See the approach in drop_exit at the end.
 */
	if (copy_arg){
		paste(&con, copy_arg, false);
	}

/* "worst" case, both input and output - fork, block-read from stdin and put to
 * output queue. cont can be shared in this corner case as the parent will still
 * be alive due to the in/out event-queue separation. */
	bool drop_exit = true;
	if (use_stdin && use_stdout){
		pid_t pv = fork();
		if (pv > 0){
			use_stdin = false;
		}
		else if (pv == 0){
			use_stdout = false;
			drop_exit = false;
		}
	}

/* block/flush stdin and split into paste- messages */
	arcan_event ev;
	if (use_stdin){
		char buf[sizeof(ev.ext.message.data) * 4];
		while(!feof(stdin) && !ferror(stdin)){
			size_t ntr = fread(buf, 1, sizeof(buf)-1, stdin);
			buf[ntr] = '\0';
			if (ntr && ntr < sizeof(buf)-1)
				paste(&con, buf, !(feof(stdin) || ferror(stdin)));
		}
	}

/* normal event loop and trigger when complete messages are received */
	if (use_stdout){
		while(arcan_shmif_wait(&con, &ev) > 0){
			if (ev.category == EVENT_TARGET){
				int sv = exec_cmd ?
					dispatch_event_outcmd(&ev, exec_cmd) :
					dispatch_event_outf(&ev, stdout);
				if (-1 == sv || (1 == sv && loop_counter == 1))
					break;
				if (loop_counter > 0)
					loop_counter--;
			}
		}
	}

/* defer the drop call until the eventqueue is empty, not a very pretty
 * solution but this is a rather fringe use-case (connect, queue one
 * event, shutdown) to play out in an asynch setting */
	if (drop_exit){
		while (con.addr && con.addr->dms &&
			con.addr->parentevq.front != con.addr->parentevq.back)
			arcan_shmif_signal(&con, SHMIF_SIGVID);
		arcan_shmif_drop(&con);
	}

	return EXIT_SUCCESS;
}
